llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp

```diff
@@ -0,0 +1,1410 @@
+      if (Node->Continue == Dst)
+        continue;
+
+      // TODO: what about cases branching to another case in the switch? Seems
+      // to work, but need to double check.
+      HasBadEdge = true;
+    }
+
+    if (!HasBadEdge)
+      return;
+
+    // Create a single exit node gathering all exit edges.
+    BasicBlock *NewExit = S.createSingleExitNode(Node->Header, Edges);
+
+    // Fixup this construct's merge node to point to the new exit.
+    // Note: this algorithm fixes inner-most divergence construct first. So
+    // recursive structures sharing a single merge node are fixed from the
+    // inside toward the outside.
+    auto MergeInstructions = getMergeInstructions(*Node->Header);
+    assert(MergeInstructions.size() == 1);
+    Instruction *I = MergeInstructions[0];
+    BlockAddress *BA = cast<BlockAddress>(I->getOperand(0));
+    if (BA->getBasicBlock() == Node->Merge) {
+      auto MergeAddress = BlockAddress::get(NewExit->getParent(), NewExit);
+      I->setOperand(0, MergeAddress);
+    }
+
+    // Clean up of the possible dangling BockAddr operands to prevent MIR
+    // comments about "address of removed block taken".
+    if (!BA->isConstantUsed())
+      BA->destroyConstant();
+
+    Node->Merge = NewExit;
+    // Regenerate the dom trees.
+    S.invalidate();
+  }
+
+  bool splitCriticalEdges(Function &F) {
+    LoopInfo &LI = getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+    Splitter S(F, LI);
+
+    DivergentConstruct Root;
+    BlockSet Visited;
+    constructDivergentConstruct(Visited, S, &*F.begin(), &Root);
+    fixupConstruct(S, &Root);
+
+    return true;
+  }
+
+  // Simplify branches when possible:
+  //  - if the 2 sides of a conditional branch are the same, transforms it to an
+  //  unconditional branch.
+  //  - if a switch has only 2 distinct successors, converts it to a conditional
+  //  branch.
+  bool simplifyBranches(Function &F) {
+    bool Modified = false;
+
+    for (BasicBlock &BB : F) {
+      SwitchInst *SI = dyn_cast<SwitchInst>(BB.getTerminator());
+      if (!SI)
+        continue;
+      if (SI->getNumCases() > 1)
+        continue;
+
+      Modified = true;
+      IRBuilder<> Builder(&BB);
+      Builder.SetInsertPoint(SI);
+
+      if (SI->getNumCases() == 0) {
+        Builder.CreateBr(SI->getDefaultDest());
+      } else {
+        Value *Condition =
+            Builder.CreateCmp(CmpInst::ICMP_EQ, SI->getCondition(),
+                              SI->case_begin()->getCaseValue());
+        Builder.CreateCondBr(Condition, SI->case_begin()->getCaseSuccessor(),
+                             SI->getDefaultDest());
+      }
+      SI->eraseFromParent();
+    }
+
+    return Modified;
+  }
+
+  // Makes sure every case target in |F| are unique. If 2 case branch to the
+  // same basic block, one of the target is updated so it jumps to a new basic
+  // block ending with a single unconditional branch to the original target.
+  bool splitSwitchCases(Function &F) {
```

Comment:
Depending on how you do this, you could get something that cannot be represented in SPIR-V:

```
switch (i) {
  case 0:
  case 1:
  case 2:
    DoSomething();
    break;
  default:
    break;
}
```

This is pretty common. I'm wondering if this funciton would turn it into:

```
switch (i) {
  case 0:
    goto case 2;
  case 1:
    goto case 2;
  case 2:
    DoSomething();
    break;
  default:
    break;
}
```

This is a problem for SPIR-V because case 0 no longer fall through to the next case. A case in spir-v can only branch to another case if it is the next one. We need to make sure that case 0 branches to case 1, and case 1 branches to case 2.

Reply:
We can deal with this in a follow up PR if something needs to be done.

Reply:
Ahh, that's not great.
But I think we should get rid of switched entirely anyway. When I was discussing with Alan, he explain that convergence is now correctly defined for if/else, but switches still have some level of under-defined convergence (fallthrough is one case). So if we want to use maximal reconvergence, and have a 100% defined behavior, I'll probably need to convert those to if/else trees.

Reply:
Mmm looks like this was indeed redundant since the simplifyBranch is done before. Removed this code, all tests are still passing.